File: leetcode/python/1167_minimum_cost_to_connect_sticks.py
import heapq
from typing import List


class Solution:

    def connectSticks(self, sticks: List[int]) -> int:
        connection_cost = 0
        heapq.heapify(sticks)
        while len(sticks) > 1:
            x, y = heapq.heappop(sticks), heapq.heappop(sticks)
            new_stick = x+y
            connection_cost += new_stick
            heapq.heappush(sticks, new_stick)
        return connection_cost

File: leetcode/java/1167MinimumCostToConnectSticks.java
/**
 * T: O(NlogN) S: O(N)
 *
 * <p>Take two smallest sticks on each iteration to minimize cost. Use priority queue to make
 * complexity logarithmic instead of square.
 */
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.stream.Collectors;

class Solution {
  public int connectSticks(int[] sticks) {
    List<Integer> sticksList = Arrays.stream(sticks).boxed().collect(Collectors.toList());
    Queue<Integer> sortedSticks = new PriorityQueue<>(sticksList);
    int totalConnectingCost = 0;
    while (sortedSticks.size() > 1) {
      int smallestStick = sortedSticks.poll();
      int secondSmallestStick = sortedSticks.poll();
      int newStick = smallestStick + secondSmallestStick;
      totalConnectingCost += newStick;
      sortedSticks.offer(newStick);
    }
    return totalConnectingCost;
  }
}
